File: C/sorting/shellSort.c
#include <math.h>
#include <stdio.h>

#define ARR_LENGTH 40

void generateKnuthSequence(int sequence[], int arrLength);

int main()
{
    int arr[ARR_LENGTH] = { 1, 7, 3, 2, 9, 5, 4, 2, 7, 3, 8, 6, 1, 7, 3, 2, 9, 5, 4, 2, 7, 3, 8, 6, 1, 7, 3, 2, 9, 5, 4, 2, 7, 3, 8, 6, 8, 4, 3, 5 };        
    int sequence[100]; // Assuming the sequence will not have more than 100 elements
    generateKnuthSequence(sequence, ARR_LENGTH);

    // Iterate over the Knuth sequence
    // and perform insertion sort for every step
    for (int stepIndex = 0; sequence[stepIndex] != 0; stepIndex++) {
        int step = sequence[stepIndex];

        for (int index = step; index < ARR_LENGTH; index++) {
            int currentItem = arr[index];
            int compareIndex = index - step;

            // Shift elements that are step distance apart and greater than currentItem
            while (compareIndex >= 0 && currentItem < arr[compareIndex]) {
                arr[compareIndex + step] = arr[compareIndex];
                compareIndex -= step;
            }

            // Place currentItem in its correct position
            arr[compareIndex + step] = currentItem;
        }
    }
}

// Generate the Knuth sequence based on the array size
// 1, 4, 13, 40, 121, ...
void generateKnuthSequence(int sequence[], int arrLength) {
    int gapSequenceLength = (int)floor(log(arrLength) / log(3)) - 1;
    int value = 1;

    for (int index = 0; index < gapSequenceLength; index++) {
        sequence[index] = value;
        value = 3 * value + 1; // Calculate the next gap size
    }

    // Reverse the sequence as per Knuth's suggestion
    for (int i = 0; i < gapSequenceLength / 2; i++) {
        int temp = sequence[i];
        sequence[i] = sequence[gapSequenceLength - i - 1];
        sequence[gapSequenceLength - i - 1] = temp;
    }

    // Marking the end of the sequence with 0
    sequence[gapSequenceLength] = 0;
}

File: rust/src/sorting/shell_sort.rs
pub fn shell_sort<T: Ord + Copy>(arr: &mut [T]) {
    if arr.is_empty() {
        return;
    }
    let sequence: Vec<i32> = generate_knuth_sequence(arr.len());

    for seq_value in sequence.iter() {
        let seq_value = *seq_value as usize; // Dereference to get the i32 value

        for index in seq_value..arr.len() {
            let current_item: T = arr[index];
            let mut compare_index: usize = index;
    
            while compare_index >= seq_value && arr[compare_index - seq_value] > current_item {
                arr[compare_index] = arr[compare_index - seq_value];
                compare_index -= seq_value;
            }
            arr[compare_index] = current_item;
        }
    }
}

// Generate the Knuth sequence based on the array size
// 1, 4, 13, 40, 121, ...
fn generate_knuth_sequence(array_size: usize) -> Vec<i32> {
    let gap_sequence_length = (f64::log(array_size as f64, 3.0)).floor() as usize - 1;

    let mut knuth_sequence: Vec<i32> = Vec::with_capacity(gap_sequence_length);
    let mut value: i32 = 1;
    knuth_sequence.push(value);

    for _index in 0..gap_sequence_length {
        value = 3 * value + 1; // Calculate the next gap size
        knuth_sequence.push(value);
    }

    knuth_sequence.reverse(); // Reverse the sequence as per Knuth's suggestion
    knuth_sequence
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty() {
        let mut empty_vec: Vec<i32> = Vec::new();
        shell_sort(&mut empty_vec);
        
        assert!(empty_vec.is_empty());
    }

    #[test]
    fn sort() {
        let mut arr: Vec<i32> = vec![4, -3, 7, 0, 10, -6, 7, 1];
        shell_sort(&mut arr);
        
        assert_eq!(arr, vec![-6, -3, 0, 1, 4, 7, 7, 10]);
    }
}
